src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -646,18 +646,135 @@ lookAhead (Parser step1 initial1 _) = Parser step initial extract
 -------------------------------------------------------------------------------
 -- Interleaving
 -------------------------------------------------------------------------------
+
+data ParserTurn = Parser1 | Parser2
+
 --
 -- | See 'Streamly.Internal.Data.Parser.deintercalate'.
 --
--- /Unimplemented/
+-- /Internal/
 --
 {-# INLINE deintercalate #-}
 deintercalate ::
-    -- Monad m =>
+    MonadCatch m =>
        Fold m a y -> Parser m x a
     -> Fold m b z -> Parser m x b
     -> Parser m x (y, z)
-deintercalate = undefined
+deintercalate
+    (Fold fstep1 finitial1 fextract1)
+    (Parser pstep1 pinitial1 pextract1)
+    (Fold fstep2 finitial2 fextract2)
+    (Parser pstep2 pinitial2 pextract2) =
+
+    Parser step initial extract
+
+    where
+
+    initial = do
+        finit1 <- finitial1
+        pinit1 <- pinitial1
+        finit2 <- finitial2
+        pinit2 <- pinitial2
+        return (finit1, pinit1, finit2, pinit2, Parser1, 0::Int)
+
+    step (fs1, ps1, fs2, ps2, currentParser, numBuffered) a =
```

Comment:
Instead of `currentParser` you can probaly simplify this if you use `Bool`. Something like `isParser1`. Your call though. You can try both and check the benchmarks. I suspect they won't differ much though.